### ISSUE TITLE
When will the produced pyi provide type information?

### ISSUE THREAD

**Issue by nschloe** [Nuitka/Nuitka#1369]
The pyi file produces by Nuitka says:
```
# In the future, this will also contain type information for values
# in the module, so IDEs will use this. Therefore please include it
# when you make software releases of the extension module that it
# describes.
```
Is there a bug report I can subscribe to for tracking this?

**Comment by kayhayen** [comment]
Now there is this one.

**Comment by kayhayen** [comment]
There is a need for more module level static optimization of class generation, then this will be doable. I expect the the Nuitka core will have the necessary bits this summer.

**Comment by nschloe** [comment]
I just found an easier way to handle this. Just run [`stubgen`](https://mypy.readthedocs.io/en/stable/stubgen.html) from mypy in the package, and move the `pyi` files into the source tree. Don't forget to include the `pyi` files in the `MANIFEST.in`, and everything works correctly.

Bottom line: I don't need the Nuitka-generated `pyi` file anymore. Is it possible to disable its generation?

**Comment by kayhayen** [comment]
Yes, it can be disabled.

**Comment by nschloe** [comment]
That's cool! Do you know if it's possible to pass those through via `pyproject.toml`'s
```toml
build-backend = "nuitka.distutils.Build"
```
?

**Comment by nschloe** [comment]
The `build` folks [say](https://github.com/pypa/build/issues/444#issuecomment-1064087997) that I'll need to wrap my own `class Build()` around Nuitka's, which I'll happily do. I just need to figure out how to set that `--no-pyi` option on the class.

**Comment by kayhayen** [comment]
I added the ability to add Nuitka options in the `pyproject.toml` earlier today. It's described in the README on factory branch, and will be in the next release of Nuitka. https://nuitka.net/doc/factory.html

I did not find any way to pass options through `setup.cfg`, they seem in fact to limit the options that can be used, so it's not expandable, but for the `pyproject.toml` with this now the `setup.py` is not needed anymore.

Nuitka has its own build class in fact in the stable release. I assume you were using it? I did not find any support for passing options to the build backend, but the build backend can read the toml file, so it was doable.

**Comment by nschloe** [comment]
> I did not find any way to pass options through setup.cfg

I found that one can run
```
python3 -m build -n --config-setting=--opt=5
```
to pass options to the `NuitkaBuildMetaBackend::build_wheel`; they are in `config_settings` there. Not sure how to pass those on to avoid the pyi file.

**Comment by kayhayen** [comment]
Well, yes, could expose Nuitka options there, but so far I am not interested. For things like `--show-scons` and `--clang` it would make sense, but generally it's not good for `--include-package-data` and the like, so I didn't yet go there, and so far I only considered it for `setup.py` where there is a method, that is pretty undocumented. Could just simply support `NUITKA_EXTRA_OPTIONS` which is used in its test runners.

Anyway, with factory you can do it now just fine.

```
   [nuitka]
   # These are not recommended, but they make it obvious to have effect.

   # boolean option, e.g. if you cared for C compilation commands, leading
   # dashes are omitted
   show-scons = true

   # options without value, e.g. enforce using Clang, empty value
   # indicates no value.
   clang = ""

   # options with single values, e.g. enable a plugin of Nuitka
   enable-plugin = pyside2

   # options with several values, e.g. avoiding including modules, accepts
   # list argument.
   nofollow-import-to = ["*.tests", "*.distutils"]
```

There is something foul with how the no value taking `show-scons` is used with boolean here, and clang is not, because I am thinking setting to false will not work. It might have to look at the `nuitka.OptionParsing` objects to work properly to set `false` there, this is all still a bit rough. And `clang = true` ought to work just fine. Well, this is all still a bit unpolished.

**Comment by nschloe** [comment]
Can you let me know which option I have to add to which file to suppress the default `.pyi` generation?

> It might have to look at the nuitka.OptionParsing objects to work properly t

Perhaps it's worthwhile switching from optparse to argparse. (optparse [is deprecated](https://docs.python.org/3/library/optparse.html) since Python 3.2.)

**Comment by kayhayen** [comment]
It has actually been deprecated since 2.7  and still it will be in 3.11, so I don't care.

As for the concrete option, I trust you can read `--help` output just as well as I can.

**Comment by nschloe** [comment]
I know of
```
nuitka3 --no-pyi-file
```
but that doesn't help me since I never run `nuitka3` directly. I build my projects with
```
build-backend = "nuitka.distutils.Build"
```
In [your comment above](https://github.com/Nuitka/Nuitka/issues/1369#issuecomment-1064882026) you quote a file, but I don't know which file that is. I also don't know what the "no-pyi" option should read there.

**Comment by kayhayen** [comment]
It is described in the user manual of factory branch, it's the `pyproject.toml` file.

**Comment by nschloe** [comment]
I finally managed to prevent Nuitka from creating the pyi file by adding
```toml
[nuitka]
no-pyi = true
```
to the project's `pyproject.toml` file (using the factory branch of Nuitka).

**Comment by kayhayen** [comment]
The issue this is about (for me) is not solved though.

**Comment by ustcfdm** [comment]
Maybe `nuitka` can just call `mypy` somehow to include type information in `pyi` file?

**Comment by kayhayen** [comment]
@ustcfdm there is actually another project, that might be better suited, but this has not seen the time to integrate it. Help is definitely welcome.

**Comment by Mase3206** [comment]
Hey, just checking in. It looks like it's been about a year since the last bit of activity. This would be a really amazing feature!

**Comment by KRRT7** [comment]
@Mase3206 Nuitka should be able to provide some simple type hinting already, try updating to latest nuitka

**Comment by kayhayen** [comment]
Part of the stable release already.

**Comment by Mase3206** [comment]
Sorry, I should have clarified. Nuitka *does* generate a single .pyi file, but it contains basically no type information and strips out docstrings. I managed to hack something together by compiling with Nuitka, unpacking the wheel file, copying over pyi files generated by `stubgen`, and repacking the wheel file. This actually allows the IDE to have type information beyond “I am a module that exists.”

**Comment by kayhayen** [comment]
You are wrong with stable Nuitka for all I know. The doc strings may indeed not be covered by our own stuben as created by @KRRT7 but I guess, that ought to be easy to add indeed, but types of functions, imports, etc. are all preserved from what I have seen. Make a new report with details about what problems you see.

**Comment by kayhayen** [comment]
Locking this, as it's now clear, this is considered answered and anything not being correct about it should be reported as a real issue and not a "no it does not" style comment.
